### PR TITLE
[ESIMD] Support assert() in ESIMD kernels

### PR DESCRIPTION
**Pull request by v-klochkov**
1) Added __assert_fail to list of functions supported in ESIMD context                                                                                                                               
2) The function __devicelib_assert_fail() must never be inlined because                                                                                                                              
   GPU RT recognizes that call is replaces/lowers it.                                                                                                                                                
   The function __assert_fail() should not be inlined because it only expands                                                                                                                        
   the code and the call of that function has nothing to do with runtime                                                                                                                             
   performance (getting there once - is the end of program)

### CODE REVIEW

File: sycl/test-e2e/ESIMD/assert.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+using namespace sycl;
+
+int main() {
+  queue Q(esimd_test::ESIMDSelector, esimd_test::createExceptionHandler());
+  esimd_test::printTestLabel(Q);
+  if (!esimd_test::isGPUDriverGE(Q, esimd_test::GPUDriverOS::LinuxAndWindows,
+                                 "26816", "101.4576")) {
+    std::cout << "Skipped. The test requires GPU driver 1.3.26816 or newer.\n";
+    // Additionally, print expected messages to pass FileCheck checks below.
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/assert.cpp:20]: FileCheck verifies that assert is not ignored and prints the correct line and users's string passed to assert() call. The wrong behavior for the test is to simply write "Test finished".

[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/test-e2e/ESIMD/assert.cpp:20]: Oh sorry I see, this path runs if we skip due to the driver, so we just print the expected string on the host to get it to pass, makes sense

**Review by sarnex — APPROVED** [review]

**Comment by v-klochkov** [comment]
@intel/dpcpp-cfe-reviewers - can you please review the 1 line fix in SemaExpr.cpp file?
It enables __assert_fail() calls in ESIMD context.

**Review by turinevgeny — APPROVED** [review]

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -257,15 +257,16 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
                              diag::err_esimd_global_in_sycl_context,
                              Sema::DeviceDiagnosticReason::Sycl);
     } else if (auto *FDecl = dyn_cast<FunctionDecl>(D)) {
-      // SYCL device function cannot be called from an ESIMD context. However,
-      // funcitons that start with '__spirv_' or '__sycl_' are exceptions to
-      // this rule.
+      // SYCL device function cannot be called from ESIMD context. However,
+      // there are some exceptions from this rule - funcitons that start
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaExpr.cpp:261]: Nit: functions*

**Review by elizabethandrews — APPROVED** [review]

File: clang/lib/Sema/SemaExpr.cpp (hunk 2 above)
Comment (v-klochkov) [reply clang/lib/Sema/SemaExpr.cpp]: Thank you!
I fixed this misprint here: https://github.com/intel/llvm/pull/9615/commits/06f1c4c31f3fb16126e9e79ebb454cd701af2555

**Comment by v-klochkov** [comment]
The fail at Windows/Graph/RecordReplay/after_use.cpp is unrelated to this PR.
It started failing after NFC fix (fix in 1 comment section). Other Windows/Graph/RecordReplay test failed _before_ NFC (basic_usm.cpp) - now it passes. Those Graph/RecordReplay/* tests seem really flaky.

**Pull request merged**